### DISCUSSION TITLE
Typst | Disable or Influence the Generation of Relative Widths

### DISCUSSION THREAD

**Discussion by thekief** [jgm/pandoc#9949] [Q&A]
With one of the recent updates the support of relative widths was added to the Typst writer. While it works fine lots of times I found more and more cases, where a rather big cell led to the other cells in this row being way too small. For example, I have the following table:

![image](https://github.com/jgm/pandoc/assets/90185239/e32f2119-3e65-45f4-939a-d9758d8b0047)

The relative width is set to:

```typst
  ...
  columns: (1.26%, 98.74%),
  ...
```

Is there a way to signal `pandoc` to disable the generation or atleast influence the written values?

**Comment by jgm** [comment]
What format are you converting from, and what is the source of the table in that format?  Hard to help otherwise.

**Comment by thekief** [comment]
I am really sorry for forgetting about an example. My source is **Markdown** and I have something like this :

```markdown
 | [x] | Check Item                                                                                                                                                  |
 | --- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |
 |     | At vero eos et accusam et justo duo dolores et ea rebum.                                                                                                    |
 |     | Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. |
 |     | At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.                 |
```

resulting in the following output:

```
#figure(
  align(center)[#table(
    columns: (1.9%, 98.1%),
    align: (auto,auto,),
    table.header([\[x\]], [Check Item],),
    table.hline(),
    [], [At vero eos et accusam et justo duo dolores et ea rebum.],
    [], [Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed
    diam nonumy eirmod tempor invidunt ut labore et dolore magna
    aliquyam erat, sed diam voluptua.],
    [], [At vero eos et accusam et justo duo dolores et ea rebum. Stet
    clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor
    sit amet.],
  )]
  , kind: table
  )
```

I looked into influencing the output uing properties [1], but besides not getting them applied properly, I suppose they only can add, rather than modify stuff


\[1\]: https://pandoc.org/typst-property-output.html

**Reply by jgm** [reply]
See the documentation on markdown pipe_tables from the User's guide:

> If any line of the Markdown source is longer than the column width, then the table will take up the full text width and the cell contents will wrap, with the relative cell widths determined by the number of dashes in the line separating the table header from the table body. (For example `---|-` would make the first column 3/4 and the second column 1/4 of the full text width.) On the other hand, if no lines are wider than column width, then cell contents will not be wrapped, and the cells will be sized to their contents.

You could set `--columns=180` (for example) and then the table won't be wider than column width and you'll get a simple table without specification of widths.

Or you could adjust the width of the separator lines to get the relative widths you desire. E.g.,
```
 | --- | ------------------------------- |
```

It is also possible to create a Lua filter that modifies the relative widths in the AST, but that's more involved.

**Reply by thekief** [reply]
Setting `--columns` somehow had no impact in my case. Using the dashes to specify the size is a nice workaround but I am not sure if it its fit for the future. The Markdown formatters I recently used, tend to automatically add the dashes as needed. 

Another example for a "broken table", which contains a bit more than the previous example, is the following example:

```
| Version        | Date       | Changes                                                                                                                                                     | Modifier |
| -------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- |
| v0.1 till v0.4 | 2023-07-06 | At vero eos et accusam et justo duo dolores et ea rebum. At vero eos et accusam et justo duo dolores et ea rebum.                                           | me       |
| v1.0           | 2023-08-28 | Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. | me       |
| v1.1           | 2024-01-25 | At vero eos et accusam et justo duo dolores et ea rebum.                                                                                                    | me       |
| v1.2           | 2024-07-09 | At vero eos et accusam et justo duo dolores et ea rebum.                                                                                                    | me       |
```

This generates the following width relation `(7.49%, 5.35%, 82.89%, 4.28%)`, resulting in the the following generated PDF

![image](https://github.com/jgm/pandoc/assets/90185239/18c9c487-d028-416e-80be-144254aba612)

If I may ask, what was the reason to pre-calculate these values? While leaving the default value may not be perfect, it is easier to detect and improve upon in, e.g. a template, than in the current case.

**Reply by jgm** [reply]
The reason is that some formats (e.g. LaTeX, Docx -- and most formats besides HTML) require widths to be specified whenever the table cells' contents do not all fit on a single line (i.e. whenever they need wrapping).  So we need to get that information in the AST.

`--columns` should work -- but note that you need it to be specified when *reading* markdown (if you just applied it in the writing phase, that's too late).

**Reply by thekief** [reply]
So, the width were added to address the issues if Typst is the source and one of these formats is the target?

As for columns `--columns`, I have the following workflow:
```
# populate the skeleton 
pp skeleton.md > _skeleton.md

# generate the Typst file
${pandoc_path} -fmarkdown --template template.typ -ttypst --toc --columns=100 -o _skeleton.typ _skeleton.md
```
If I understand you correcty, I would need to convert `Markdown` to `Markdown` beforehand to apply it, right?

As for my issue, would it be feasible to use the Typst properties to overwrite the generated width (in the future)? While stripping the current width's, search- & replacing them, etc. works, I would like to specify it directly. 

And I gotta be honest, I know where to apply certain properties to, but I am not sure where the ones for tables need to be added. So potentially the behaviour alread exists, but I couldn't apply it. Could you give me an example, where I would, e.g. add `
{.typst:table:align="(auto,auto)"}` to a Markdown table to change the alignment?

**Reply by jgm** [reply]
That should work, maybe try with `--columns=999` (maybe your table in `_skeleton.md` is wider than 100 characters).

To attach typst properties to a table, you'd need to use a Lua filter, because there's no way to specify table attributes directly in Markdown.  Not sure even then if this would work (without checking source), but you could try.
